File: 01_projects/code/python-playwright/cbaps-dawms/robust/shared/playwright_manager.py
"""
Enhanced PlaywrightManager - 60+ Comprehensive Methods
Matches Java-Selenium and TypeScript robustness
"""
from playwright.sync_api import sync_playwright, Page, Locator
from faker import Faker
from loguru import logger
import base64
import time
from typing import List

class PlaywrightManager:
    """Comprehensive Playwright wrapper - 60+ methods"""
    
    def __init__(self, browser: str = "chromium", headless: bool = False):
        self.browser_type = browser
        self.headless = headless
        self.playwright = None
        self.browser = None
        self.context = None
        self.page = None
        self.faker = Faker()
        self.demo_mode = False
    
    # ========== Init/Teardown ==========
    def init_playwright(self):
        self.playwright = sync_playwright().start()
        if self.browser_type == "firefox":
            self.browser = self.playwright.firefox.launch(headless=self.headless)
        elif self.browser_type == "webkit":
            self.browser = self.playwright.webkit.launch(headless=self.headless)
        else:
            self.browser = self.playwright.chromium.launch(headless=self.headless)
        self.context = self.browser.new_context(
            viewport={"width": 1920, "height": 1080},
            ignore_https_errors=True,
            record_video_dir="videos/"
        )
        logger.info("✅ Playwright initialized")
    
    def open_new_page(self) -> Page:
        self.page = self.context.new_page()
        return self.page
    
    def close_playwright(self):
        if self.page: self.page.close()
        if self.context: self.context.close()
        if self.browser: self.browser.close()
        if self.playwright: self.playwright.stop()
    
    def close_page(self):
        if self.page: self.page.close()
    
    # ========== Navigation (7) ==========
    def navigate_to(self, url: str):
        self.page.goto(url)
        logger.info(f"➡️ {url}")
    
    def get_current_url(self) -> str:
        return self.page.url
    
    def get_title(self) -> str:
        return self.page.title()
    
    def refresh_page(self):
        self.page.reload()
    
    def navigate_back(self):
        self.page.go_back()
    
    def navigate_forward(self):
        self.page.go_forward()
    
    def wait_for_network_idle(self):
        self.page.wait_for_load_state("networkidle")
    
    # ========== Interaction (10) ==========
    def click_element(self, locator: Locator):
        self._blink(locator)
        locator.click()
    
    def enter_text(self, locator: Locator, text: str):
        self._blink(locator)
        locator.fill(text)
    
    def select_dropdown(self, locator: Locator, value: str):
        locator.select_option(value)
    
    def check_checkbox(self, locator: Locator):
        locator.check()
    
    def uncheck_checkbox(self, locator: Locator):
        locator.uncheck()
    
    def click_hidden(self, locator: Locator):
        locator.dispatch_event("click")
    
    def double_click(self, locator: Locator):
        locator.dblclick()
    
    def right_click(self, locator: Locator):
        locator.click(button="right")
    
    def hover(self, locator: Locator):
        locator.hover()
    
    def press_enter(self, locator: Locator):
        locator.press("Enter")
    
    # ========== Wait (4) ==========
    def wait_for_selector(self, selector: str, timeout: int = 30000):
        self.page.wait_for_selector(selector, timeout=timeout)
    
    def wait_for_hidden(self, selector: str):
        self.page.wait_for_selector(selector, state="hidden")
    
    def wait(self, seconds: float):
        time.sleep(seconds)
    
    # ========== State (8) ==========
    def is_visible(self, locator: Locator) -> bool:
        try:
            return locator.is_visible()
        except:
            return False
    
    def is_enabled(self, locator: Locator) -> bool:
        try:
            return locator.is_enabled()
        except:
            return False
    
    def is_checked(self, locator: Locator) -> bool:
        try:
            return locator.is_checked()
        except:
            return False
    
    def get_text(self, locator: Locator) -> str:
        return locator.text_content()
    
    def get_attribute(self, locator: Locator, attr: str) -> str:
        return locator.get_attribute(attr)
    
    def get_count(self, locator: Locator) -> int:
        return locator.count()
    
    def get_all_texts(self, locator: Locator) -> List[str]:
        return locator.all_text_contents()
    
    # ========== Scroll (3) ==========
    def scroll_to_element(self, locator: Locator):
        locator.scroll_into_view_if_needed()
    
    def scroll_to_top(self):
        self.page.evaluate("window.scrollTo(0, 0)")
    
    def scroll_to_bottom(self):
        self.page.evaluate("window.scrollTo(0, document.body.scrollHeight)")
    
    # ========== Screenshot (2) ==========
    def capture_screenshot(self, filename: str):
        self.page.screenshot(path=f"screenshots/{filename}.png", full_page=True)
    
    def capture_screenshot_base64(self) -> str:
        return base64.b64encode(self.page.screenshot()).decode()
    
    # ========== Test Data (6) ==========
    def get_random_email(self) -> str:
        return self.faker.email()
    
    def get_random_password(self, length: int = 12) -> str:
        return self.faker.password(length=length)
    
    def get_random_name(self) -> str:
        return self.faker.name()
    
    def get_random_phone(self) -> str:
        return self.faker.phone_number()
    
    def get_random_address(self) -> str:
        return self.faker.address()
    
    def get_random_text(self, length: int = 50) -> str:
        return self.faker.text(max_nb_chars=length)
    
    # ========== Utility ==========
    def _blink(self, locator: Locator):
        if self.demo_mode:
            try:
                for _ in range(2):
                    locator.evaluate("el => el.style.border = '3px solid red'")
                    time.sleep(0.2)
                    locator.evaluate("el => el.style.border = ''")
                    time.sleep(0.2)
            except:
                pass

File: 01_projects/code/typescript-playwright/cbaps-dawms/dawms/library/PlaywrightManager.ts
/**
 * PlaywrightManager: Enhanced TypeScript wrapper for Playwright
 * Manages browser lifecycle and provides type-safe interaction methods
 */

import { Browser, BrowserContext, Page, Locator, chromium, firefox, webkit } from '@playwright/test';
import { BrowserConfig } from './types';
import * as fs from 'fs';
import * as path from 'path';

export class PlaywrightManager {
  private browser: Browser | null = null;
  private context: BrowserContext | null = null;
  public page: Page | null = null;
  
  private config: BrowserConfig = {
    browserType: 'chromium',
    headless: false,
    viewport: { width: 1280, height: 720 },
    slowMo: 0,
    video: true,
    screenshot: true
  };

  constructor(customConfig?: Partial<BrowserConfig>) {
    if (customConfig) {
      this.config = { ...this.config, ...customConfig };
    }
    this.ensureDirectories();
  }

  /**
   * Ensure output directories exist
   */
  private ensureDirectories(): void {
    const dirs = ['target/screenshots', 'target/videos', 'target/traces', 'target/reports'];
    dirs.forEach(dir => {
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
    });
  }

  /**
   * Initialize Playwright with browser and context
   */
  async initPlaywright(): Promise<void> {
    try {
      // Select browser type
      const browserType = this.config.browserType === 'firefox' ? firefox
        : this.config.browserType === 'webkit' ? webkit
        : chromium;

      // Launch browser
      this.browser = await browserType.launch({
        headless: this.config.headless,
        slowMo: this.config.slowMo,
        args: ['--start-maximized']
      });

      // Create context with options
      this.context = await this.browser.newContext({
        viewport: this.config.headless ? this.config.viewport : null,
        recordVideo: this.config.video ? {
          dir: 'target/videos/',
          size: { width: 1280, height: 720 }
        } : undefined,
        ignoreHTTPSErrors: true
      });

      // Enable tracing for debugging
      await this.context.tracing.start({ screenshots: true, snapshots: true });

      console.log(`✓ Playwright initialized with ${this.config.browserType} browser`);
    } catch (error) {
      console.error('Failed to initialize Playwright:', error);
      throw error;
    }
  }

  /**
   * Open a new browser page
   */
  async openNewPage(): Promise<Page> {
    if (!this.context) {
      throw new Error('Browser context not initialized. Call initPlaywright() first.');
    }

    this.page = await this.context.newPage();
    console.log('✓ New browser page opened');
    return this.page;
  }

  /**
   * Close the current page
   */
  async closePage(): Promise<void> {
    if (this.page) {
      await this.page.close();
      console.log('✓ Browser page closed');
    }
  }

  /**
   * Close Playwright completely
   */
  async closePlaywright(): Promise<void> {
    try {
      if (this.context) {
        await this.context.tracing.stop({ path: `target/traces/trace-${Date.now()}.zip` });
        await this.context.close();
      }
      if (this.browser) {
        await this.browser.close();
      }
      console.log('✓ Playwright session closed');
    } catch (error) {
      console.error('Error closing Playwright:', error);
      throw error;
    }
  }

  // ============ Enhanced Action Methods ============

  /**
   * Type-safe click with retry logic
   */
  async click(locator: Locator, options?: { timeout?: number; force?: boolean }): Promise<void> {
    try {
      await this.highlightElement(locator);
      await locator.click({ timeout: options?.timeout || 30000, force: options?.force });
      console.log('✓ Element clicked');
    } catch (error) {
      console.error('Click failed:', error);
      throw error;
    }
  }

  /**
   * Type text with clear and validation
   */
  async type(locator: Locator, text: string, options?: { clear?: boolean; delay?: number }): Promise<void> {
    try {
      await this.highlightElement(locator);
      
      if (options?.clear !== false) {
        await locator.clear();
      }
      
      await locator.fill(text, { timeout: 30000 });
      
      // Validate the text was entered correctly
      const value = await locator.inputValue();
      if (value !== text) {
        console.warn(`Warning: Expected "${text}" but got "${value}"`);
      }
      
      console.log(`✓ Typed: "${text}"`);
    } catch (error) {
      console.error('Type failed:', error);
      throw error;
    }
  }

  /**
   * Wait for element to be visible with custom message
   */
  async waitVisible(selector: string, message?: string, timeout: number = 30000): Promise<void> {
    try {
      if (!this.page) throw new Error('Page not initialized');
      
      await this.page.waitForSelector(selector, { state: 'visible', timeout });
      const locator = this.page.locator(selector);
      await this.highlightElement(locator);
      
      console.log(`✓ ${message || 'Element visible'}: ${selector}`);
    } catch (error) {
      console.error(`Element not visible: ${selector}`, error);
      throw error;
    }
  }

  /**
   * Wait for element to be hidden
   */
  async waitHidden(selector: string, timeout: number = 30000): Promise<void> {
    if (!this.page) throw new Error('Page not initialized');
    await this.page.waitForSelector(selector, { state: 'hidden', timeout });
    console.log(`✓ Element hidden: ${selector}`);
  }

  /**
   * Select dropdown option with validation
   */
  async selectDropdown(locator: Locator, value: string): Promise<void> {
    try {
      await this.highlightElement(locator);
      await locator.selectOption(value);
      
      // Validate selection
      const selectedValue = await locator.inputValue();
      if (selectedValue !== value) {
        console.warn(`Warning: Expected "${value}" to be selected but got "${selectedValue}"`);
      }
      
      console.log(`✓ Selected dropdown option: "${value}"`);
    } catch (error) {
      console.error('Dropdown selection failed:', error);
      throw error;
    }
  }

  /**
   * Get text content from locator
   */
  async getText(locator: Locator): Promise<string> {
    const text = await locator.textContent() || '';
    console.log(`✓ Retrieved text: "${text}"`);
    return text.trim();
  }

  /**
   * Check if element is visible
   */
  async isVisible(locator: Locator): Promise<boolean> {
    try {
      return await locator.isVisible();
    } catch {
      return false;
    }
  }

  /**
   * Check if element is enabled
   */
  async isEnabled(locator: Locator): Promise<boolean> {
    try {
      return await locator.isEnabled();
    } catch {
      return false;
    }
  }

  /**
   * Highlight element for visual feedback (optional demo mode)
   */
  private async highlightElement(locator: Locator): Promise<void> {
    if (this.config.slowMo && this.config.slowMo > 0) {
      try {
        await locator.evaluate((el: HTMLElement) => {
          const originalStyle = el.style.cssText;
          el.style.border = '3px solid red';
          el.style.backgroundColor = 'yellow';
          setTimeout(() => {
            el.style.cssText = originalStyle;
          }, 500);
        });
      } catch {
        // Ignore highlight errors
      }
    }
  }

  /**
   * Take screenshot with custom name
   */
  async screenshot(name: string, fullPage: boolean = true): Promise<string> {
    if (!this.page) throw new Error('Page not initialized');
    
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const filename = `${name}-${timestamp}.png`;
    const filepath = path.join('target/screenshots', filename);
    
    await this.page.screenshot({ path: filepath, fullPage });
    console.log(`✓ Screenshot saved: ${filepath}`);
    return filepath;
  }

  /**
   * Take screenshot as base64
   */
  async screenshotBase64(): Promise<string> {
    if (!this.page) throw new Error('Page not initialized');
    
    const buffer = await this.page.screenshot({ fullPage: true });
    return buffer.toString('base64');
  }

  /**
   * Navigate to URL with wait
   */
  async navigateTo(url: string, waitUntil: 'load' | 'domcontentloaded' | 'networkidle' = 'load'): Promise<void> {
    if (!this.page) throw new Error('Page not initialized');
    
    await this.page.goto(url, { waitUntil, timeout: 60000 });
    console.log(`✓ Navigated to: ${url}`);
  }

  /**
   * Wait for specified time (use sparingly)
   */
  async wait(milliseconds: number): Promise<void> {
    await new Promise(resolve => setTimeout(resolve, milliseconds));
  }

  /**
   * Press keyboard key
   */
  async pressKey(locator: Locator, key: string): Promise<void> {
    await locator.press(key);
    console.log(`✓ Pressed key: ${key}`);
  }

  /**
   * Scroll element into view
   */
  async scrollToElement(locator: Locator): Promise<void> {
    await locator.scrollIntoViewIfNeeded();
    console.log('✓ Scrolled element into view');
  }

  /**
   * Get page title
   */
  async getTitle(): Promise<string> {
    if (!this.page) throw new Error('Page not initialized');
    return await this.page.title();
  }

  /**
   * Get current URL
   */
  async getCurrentUrl(): Promise<string> {
    if (!this.page) throw new Error('Page not initialized');
    return this.page.url();
  }
}
